File: fann.c
#include <stdio.h>
#include <stdio.h>

int	res2;

static int fann1(int n) {
int	signx;
int	maxflips;
int	sum;
int	i;
int	j;
int	k;
int	q1;
int	flips;
int	qq;
int	t;
int	sx;
int	tt;
int	p[100];
int	q[100];
int	s[100];
    signx = 1;
    maxflips = 0;
    sum = 0;
    for (i=1; i<=n; ++i) {
        p[i-1] = i;
        q[i-1] = i;
        s[i-1] = i;
    }
    while (1) {
        q1 = p[1-1];
        if (q1 != 1) {
            for (i=2; i<=n; ++i) {
                q[i-1] = p[i-1];
            }
            flips = 1;
            while (1) {
                qq = q[q1-1];
                if (qq == 1) {
                    sum += signx*flips;
                    if (flips > maxflips) {
                        maxflips = flips;
                    }
                    goto L12;
                }
                q[q1-1] = q1;
                if (q1 >= 4) {
                    i = 2;
                    j = q1-1;
                    do {
                        {int	temp;
                            temp = q[i-1];
                            q[i-1] = q[j-1];
                            q[j-1] = temp;
                        }
                        ++i;
                        --j;
                    } while (!(i >= j));
                }
                q1 = qq;
                ++flips;
            }
L12:;
        }
        if (signx == 1) {
            {int	temp;
                temp = p[2-1];
                p[2-1] = p[1-1];
                p[1-1] = temp;
            }
            signx = -1;
        }
        else {
            {int	temp;
                temp = p[2-1];
                p[2-1] = p[3-1];
                p[3-1] = temp;
            }
            signx = 1;
            for (i=3; i<=n; ++i) {
                sx = s[i-1];
                if (sx != 1) {
                    s[i-1] = sx-1;
                    goto L19;
                }
                if (i == n) {
                    res2 = maxflips;
                    return sum;
                }
                s[i-1] = i;
                tt = p[1-1];
                for (j=1; j<=i; ++j) {
                    p[j-1] = p[j+1-1];
                }
                p[i+1-1] = tt;
            }
L19:;
        }
    }
    return 0;
}

int main(void) {
    int res;
    enum {n=10};
    res = fann1(n);
    printf("fannkuch(%d) = %d %d\n",n,res,res2);
}

File: fann.zig
const print = @import("std").debug.print;

var res2:i32=0;

pub fn fann(n:usize) i32 {
    var p=[20]usize {0,0,0,0,0, 0,0,0,0,0, 0,0,0,0,0, 0,0,0,0,0};
    var q=[20]usize {0,0,0,0,0, 0,0,0,0,0, 0,0,0,0,0, 0,0,0,0,0};
    var s=[20]usize {0,0,0,0,0, 0,0,0,0,0, 0,0,0,0,0, 0,0,0,0,0};

    var sign:i32=1;
    var maxflips:i32=0;
    var sum:i32=0;
    var i:usize=0;
    var j:usize=0;
    var q1:usize=0;
    var flips:i32=0;
    var qq:usize=0;
    var t:usize=0;
    var tt:usize=0;
    var sx:usize=0;

    i=1;
    while (i<=n) {
        p[i]=i;
        q[i]=i;
        s[i]=i;
        i+=1;
    }

    while (true) {
        q1=p[1];
        if (q1!=1) {
            i=2;
            while (i<=n) { q[i]=p[i]; i=i+1;}
            flips=1;
            while (true) {
                qq=q[q1];
                if (qq==1) {
                    sum+=sign*flips;
                    if (flips>maxflips) {
                        maxflips=flips;
                    }
                    break;
                }
                q[q1]=q1;
                if (q1>=4) {
                    i=2; j=q1-1;
                    while (true) {
                        t=q[i]; q[i]=q[j]; q[j]=t;
                        i+=1;
                        j-=1;
                        if (i>=j) {break;}
                    }
                }
                q1=qq;
                flips+=1;
            }
        }

        if (sign==1) {
            t=p[1]; p[1]=p[2]; p[2]=t;
            sign=-1;
        } else {
            t=p[2]; p[2]=p[3]; p[3]=t;
            sign=1;
            i=3;
            while (i<=n) {
                sx=s[i];
                if (sx!=1) { s[i]=sx-1; break; }
                if (i==n) {
                    res2=maxflips;
                    return sum;
                }
                s[i]=i;
                tt=p[1];
                j=1;
                while (j<=i) {
                    p[j]=p[j+1]; j+=1;
                }
                p[i+1]=tt;
                i+=1;
            }
        }
    }

    return 0;
}

pub fn main() !void {
    var n: usize = 11;
    var res: i32=0;

    res=fann(n);

    print("{d} ", .{res});
    print("{d}\n", .{res2});
}
